parser/src/aspects/call_parser.rs

```diff
@@ -1,37 +1,223 @@
+                    .is_some()
+                {
+                    RedirOp::String
+                } else {
+                    RedirOp::Read
+                }
+            }
+            TokenType::Greater => match self.cursor.advance(of_type(TokenType::Greater)) {
+                None => RedirOp::Write,
+                Some(_) => RedirOp::Append,
+            },
+            _ => Err(self.mk_parse_error("Expected redirection operator."))?,
+        };
+
+        // Parse file descriptor duplication and update the operator
+        if self.cursor.advance(of_type(TokenType::Ampersand)).is_some() {
+            operator = match operator {
+                RedirOp::Read => RedirOp::FdIn,
+                RedirOp::Write => RedirOp::FdOut,
+                _ => Err(self.mk_parse_error("Invalid redirection operator."))?,
+            };
+        }
+
+        let operand = self.expression()?;
+        Ok(Redir {
+            fd,
+            operator,
+            operand,
+        })
+    }
+
+    fn redirectable(&mut self, expr: Expr<'a>) -> ParseResult<Expr<'a>> {
+        let mut redirections = vec![];
+        self.cursor.advance(spaces());
+        while self.cursor.lookahead(eox()).is_none() {
+            match self.cursor.peek().token_type {
+                TokenType::Ampersand | TokenType::Less | TokenType::Greater => {
+                    redirections.push(self.redirection()?);
+                }
+                TokenType::Pipe => {
+                    return self.pipeline(expr);
+                }
+                // Detect redirections without a specific file descriptor
+                _ if self
+                    .cursor
+                    .lookahead(next().then(of_types(&[TokenType::Less, TokenType::Greater])))
+                    .is_some() =>
+                {
+                    redirections.push(self.redirection()?)
+                }
+                _ => break,
+            };
+            self.cursor.advance(spaces());
+        }
+        if redirections.is_empty() {
+            Ok(expr)
+        } else {
+            Ok(Expr::Redirected(Redirected {
+                expr: Box::new(expr),
+                redirections,
+            }))
+        }
     }
 }
 
 #[cfg(test)]
 mod tests {
+    use crate::aspects::call_parser::CallParser;
+    use crate::ast::callable::{Call, Redir, RedirFd, RedirOp, Redirected};
+    use crate::ast::literal::Literal;
+    use crate::ast::Expr;
+    use crate::parse;
+    use crate::parser::Parser;
     use lexer::lexer::lex;
     use lexer::token::{Token, TokenType};
 
-    use crate::ast::callable::Call;
-    use crate::ast::literal::{Literal, LiteralValue};
-    use crate::ast::Expr;
+    #[test]
+    fn redirection() {
+        let tokens = vec![
+            Token::new(TokenType::Identifier, "ls"),
+            Token::new(TokenType::Greater, ">"),
+            Token::new(TokenType::Identifier, "/tmp/out"),
+        ];
```

Comment:
For your tests, could you please use the lexer instead of hand-writing the token sequences ? for example in this test, i doubt that the lexer would lexe `/tmp/out` as an identifier. If no, i think it still better to rely on the lexer in order to test the compatibility.

Reply:
You're absolutely right. :p